### PR TITLE
Add statistic endpoint

### PR DESCRIPTION
**Pull request by Aaron2000119**
Closes https://github.com/Gamify-IT/issues/issues/381.

Adds statistic endpoints for the lecturer-interface.

### CODE REVIEW

File: src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+            // iterate over all correct answered questions and add correct answered counter for this problematic question
+            countRightAnsweredQuestions(problematicQuestions, gameResult);
+        }
+        sortProblematicQuestionsByPercentageWrongAnswers(problematicQuestions);
+        return problematicQuestions.subList(0, Math.min(MAX_PROBLEMATIC_QUESTIONS, problematicQuestions.size()));
+    }
+
+    /**
+     * Counts amount of right answered questions and adds them to the problematic question list
```
Comment (maexled) [inline src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java:63]:
```suggestion
     * Counts amount of right answered questions and adds them to the problematic question counter
```

File: src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                    .filter(problematicQuestion ->
+                        problematicQuestion.getQuestion().getQuestionText().equals(gameAnswer.getQuestion())
+                    )
+                    .findAny()
+                    .ifPresent(ProblematicQuestion::addCorrectAnswer)
+            );
+    }
+
+    /**
+     * Counts amount of wrong answered questions and adds them to the problematic question list
```
Comment (maexled) [inline src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java:85]:
```suggestion
     * Counts amount of wrong answered questions and adds them to the problematic question counter
```

File: src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+     *
+     * @param problematicQuestions the list of problematic questions where the wrong answered counter should be increased
+     * @param gameResult the game result which contains the answered questions
+     */
+    private void countWrongAnsweredQuestions(List<ProblematicQuestion> problematicQuestions, GameResult gameResult) {
```
Comment (maexled) [inline src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java:90]:
```suggestion
    private void countWrongAnsweredQuestions(final List<ProblematicQuestion> problematicQuestions, final GameResult gameResult) {
```

File: src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                    .ifPresent(ProblematicQuestion::addWrongAnswer)
+            );
+    }
+
+    /**
+     * Sorts the list of problematic questions by the amount of wrong answers to attempts
+     *
+     * @param problematicQuestions the list of problematic questions to sort
+     */
+    private void sortProblematicQuestionsByPercentageWrongAnswers(List<ProblematicQuestion> problematicQuestions) {
```
Comment (maexled) [inline src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java:111]:
```suggestion
    private void sortProblematicQuestionsByPercentageWrongAnswers(final List<ProblematicQuestion> problematicQuestions) {
```

File: src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        problematicQuestions.sort((o1, o2) -> {
+            double percantageWrong1 = (double) o1.getWrongAnswers() / (double) o1.getAttempts();
+            double percantageWrong2 = (double) o2.getWrongAnswers() / (double) o2.getAttempts();
```
Comment (maexled) [inline src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java:114]:
```suggestion
            final double percantageWrong1 = (double) o1.getWrongAnswers() / (double) o1.getAttempts();
            final double percantageWrong2 = (double) o2.getWrongAnswers() / (double) o2.getAttempts();
```

File: src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            timeSpentDistribution.setFromPercentage(TIME_SPENT_DISTRIBUTION_PERCENTAGES[i]);
+            timeSpentDistribution.setToPercentage(TIME_SPENT_DISTRIBUTION_PERCENTAGES[i + 1]);
+            timeSpentDistributions.add(timeSpentDistribution);
+        }
+        // order game results by time spent
+        sortGameResultsByPlayedTime(gameResults);
+
+        // calculate time spent time borders to time spent distribution percentage
+        int currentGameResultIndex = 0;
+        for (TimeSpentDistribution timeSpentDistribution : timeSpentDistributions) {
```
Comment (maexled) [inline src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java:145]:
```suggestion
        for (final TimeSpentDistribution timeSpentDistribution : timeSpentDistributions) {
```

File: src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        }
+        return timeSpentDistributions;
+    }
+
+    /**
+     * Sorts a list of game results by played time
+     *
+     * @param gameResults the list of game results to sort
+     */
+    private void sortGameResultsByPlayedTime(List<GameResult> gameResults) {
```
Comment (maexled) [inline src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java:167]:
```suggestion
    private void sortGameResultsByPlayedTime(final List<GameResult> gameResults) {
```

File: src/test/java/de/unistuttgart/crosswordbackend/statistic/StatisticTest.java [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+
+        randomConfiguration = new Configuration();
+        randomConfiguration.setQuestions(questions);
+        randomConfiguration.setName("randomConfiguration");
+        randomConfiguration = configurationRepository.save(randomConfiguration);
+
+        gameResults = new ArrayList<>();
+
+        for (int i = 0; i < 100; i++) {
+            GameResultDTO gameResultDTO = new GameResultDTO();
```
Comment (maexled) [inline src/test/java/de/unistuttgart/crosswordbackend/statistic/StatisticTest.java:106]:
```suggestion
            final GameResultDTO gameResultDTO = new GameResultDTO();
```

File: src/test/java/de/unistuttgart/crosswordbackend/statistic/StatisticTest.java [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+    private QuestionDTO problematicQuestion;
+    private QuestionDTO bestAnsweredQuestion;
+    private List<GameResult> gameResults;
+
+    @BeforeEach
+    public void createBasicData() {
+        gameResultRepository.deleteAll();
+        configurationRepository.deleteAll();
+
+        Set<Question> questions = new HashSet<>();
```
Comment (maexled) [inline src/test/java/de/unistuttgart/crosswordbackend/statistic/StatisticTest.java:93]:
```suggestion
        final Set<Question> questions = new HashSet<>();
```

**Review by maexled — APPROVED** [review]
Missed some `final`s, especially in the StatisticTest. Overall lgtm

File: src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * Returns a list of the most problematic questions of a minigame
+     *
+     * @param configurationId the configuration id of the minigame
+     * @return a list of the most problematic questions of a minigame
+     */
+    public List<ProblematicQuestion> getProblematicQuestions(final UUID configurationId) {
+        final Configuration configuration = configService.getConfiguration(configurationId);
+        final List<GameResult> gameResults = gameResultRepository.findByConfiguration(configurationId);
+        final List<ProblematicQuestion> problematicQuestions = new ArrayList<>();
```
Comment (delvh) [inline src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java:46]:
```suggestion
        final List<ProblematicQuestion> problematicQuestions = new ArrayList<>(configuration.getQuestions().size());
```

File: src/main/java/de/unistuttgart/crosswordbackend/data/statistic/ProblematicQuestion.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+@NoArgsConstructor
+@AllArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class ProblematicQuestion {
+
+    int attempts;
+    int correctAnswers;
+    int wrongAnswers;
+
+    QuestionDTO question;
```
Comment (delvh) [inline src/main/java/de/unistuttgart/crosswordbackend/data/statistic/ProblematicQuestion.java:20]:
```suggestion
    final QuestionDTO question;
```

File: src/main/java/de/unistuttgart/crosswordbackend/data/statistic/ProblematicQuestion.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.experimental.FieldDefaults;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class ProblematicQuestion {
```
Comment (delvh) [inline src/main/java/de/unistuttgart/crosswordbackend/data/statistic/ProblematicQuestion.java:14]: I think the `AllArgsConstructor` is wanted here.
You only want the constructor
```java
public ProblematicQuestion(QuestionDTO question) {
  this.question = question;
}
```

File: src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        });
+    }
+
+    /**
+     * Returns a list of the time spent distribution of a minigame
+     *
+     * @param configurationId the configuration id of the minigame
+     * @return a list of the time spent distribution of a minigame
+     */
+    public List<TimeSpentDistribution> getTimeSpentDistributions(final UUID configurationId) {
```
Comment (delvh) [inline src/main/java/de/unistuttgart/crosswordbackend/service/StatisticService.java:133]: While your hardcoded approach probably works, we probably could have simplified things a lot using a fitting SQL query.
But it is what it is.

**Review by delvh — APPROVED** [review]

**Pull request merged**